local_planner/src/nodes/waypoint_generator.cpp

```diff
@@ -208,40 +208,31 @@ void WaypointGenerator::reachGoalAltitudeFirst() {
   pose_to_wp *= planner_info_.min_speed;
 
   output_.goto_position = toPoint(toEigen(pose_.pose.position) + pose_to_wp);
+  output_.adapted_goto_position = output_.goto_position;
+  output_.smoothed_goto_position = output_.goto_position;
+  smoothed_goto_location_ = toEigen(output_.smoothed_goto_position);
 }
 
 void WaypointGenerator::smoothWaypoint(double dt) {
-  Eigen::Vector2f vel_cur_xy(curr_vel_.twist.linear.x,
-                             curr_vel_.twist.linear.y);
-  Eigen::Vector2f vel_sp_xy((output_.adapted_goto_position.x -
-                             last_position_waypoint_.pose.position.x) /
-                                dt,
-                            (output_.adapted_goto_position.y -
-                             last_position_waypoint_.pose.position.y) /
-                                dt);
-
-  Eigen::Vector2f accel_diff = (vel_sp_xy - vel_cur_xy) / dt;
-  Eigen::Vector2f accel_cur = (vel_cur_xy - last_velocity_) / dt;
-  Eigen::Vector2f jerk_diff = (accel_diff - accel_cur) / dt;
-  float max_jerk = max_jerk_limit_param_;
-
-  // velocity-dependent max jerk
-  if (min_jerk_limit_param_ > 0.001f) {
-    max_jerk *= vel_cur_xy.norm();
-    if (max_jerk < min_jerk_limit_param_) max_jerk = min_jerk_limit_param_;
-  }
-
-  if (jerk_diff.squaredNorm() > max_jerk * max_jerk && max_jerk > 0.001f) {
-    jerk_diff = max_jerk * jerk_diff.normalized();
-    vel_sp_xy = (jerk_diff * dt + accel_cur) * dt + vel_cur_xy;
+  if (smoothing_speed_ > 0.01) {
+    const float P_constant = smoothing_speed_;
+    const float D_constant = 2.f * std::sqrt(P_constant);  // critically damped
+    const Eigen::Vector3f desired_location =
+        toEigen(output_.adapted_goto_position);
+    const Eigen::Vector3f desired_velocity = toEigen(curr_vel_.twist.linear);
+
+    const Eigen::Vector3f p =
+        (desired_location - smoothed_goto_location_) * P_constant;
+    const Eigen::Vector3f d =
+        (desired_velocity - smoothed_goto_velocity_) * D_constant;
+
+    smoothed_goto_velocity_ += (p + d) * dt;
+    smoothed_goto_location_ += smoothed_goto_velocity_ * dt;
+
+    output_.smoothed_goto_position = toPoint(smoothed_goto_location_);
```

Comment:
It might involve some juggling around with geometry_msgs and Eigen vectors, but could we extend this logic to also smooth the setpoint in orientation (yaw) and/or yaw rates? Most of the jerky behavior that makes for a bad UX comes from twitches in yaw I think

Reply:
Agreed, the yaw angular rates are still similarly noisy to the old one, while the pitch and roll improved a lot. Unfortunately I don't think there is much we can do at the smoothing side without impacting the performance of avoiding obstacles - since most of the actual signal in the unsmoothed setpoint movement is in yaw, that is where most of the noise still is. I think the biggest help here will be getting smoother outputs from the planner - a smoother can only do so much.

Reply:
Yaw setpoint wasn't using the smoothed setpoint! Big improvement in yaw angular rates:
https://logs.px4.io/plot_app?log=fbc84d4f-c683-4582-bea5-1cc4c623bdaa